### PR TITLE
Fix windows file limit check

### LINKED ISSUE #10054: 2.10.17 will not start on windows host
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

PR #10014 introduced a dependency on the resource module to get the soft file limit, , which is *nix only. 

This causes ```prefect server start``` to fail on windows. (below cap is from me grabbing source and running ```pip install -e .```

### Reproduction

```python3
just run prefect server start, prefect version, pretty much anything on a windows box
```

### Error

```python3
(py311) PS C:\temp\prefect> prefect server start
Traceback (most recent call last):
  File "c:\envs\py311\Scripts\prefect-script.py", line 33, in <module>
    sys.exit(load_entry_point('prefect', 'console_scripts', 'prefect')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\envs\py311\Scripts\prefect-script.py", line 25, in importlib_load_entry_point
    return next(matches).load()
           ^^^^^^^^^^^^^^^^^^^^
  File "c:\envs\py311\Lib\importlib\metadata\__init__.py", line 202, in load
    module = import_module(match.group('module'))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\envs\py311\Lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1128, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "c:\temp\prefect\src\prefect\__init__.py", line 37, in <module>
    from prefect.states import State
  File "c:\temp\prefect\src\prefect\states.py", line 14, in <module>
    from prefect.client.schemas import State as State
  File "c:\temp\prefect\src\prefect\client\schemas\__init__.py", line 4, in <module>
    from .actions import BlockTypeUpdate, StateCreate
  File "c:\temp\prefect\src\prefect\client\schemas\actions.py", line 9, in <module>
    import prefect.client.schemas.objects as objects
  File "c:\temp\prefect\src\prefect\client\schemas\objects.py", line 20, in <module>
    from prefect._internal.schemas.bases import ObjectBaseModel, PrefectBaseModel
  File "c:\temp\prefect\src\prefect\_internal\schemas\bases.py", line 18, in <module>
    from prefect._internal.compatibility.experimental import experiment_enabled
  File "c:\temp\prefect\src\prefect\_internal\compatibility\experimental.py", line 19, in <module>
    from prefect.settings import PREFECT_EXPERIMENTAL_WARN, SETTING_VARIABLES, Setting
  File "c:\temp\prefect\src\prefect\settings.py", line 70, in <module>
    from prefect._internal.compatibility.deprecated import generate_deprecation_message
  File "c:\temp\prefect\src\prefect\_internal\compatibility\deprecated.py", line 21, in <module>
    from prefect.utilities.importtools import (
  File "c:\temp\prefect\src\prefect\utilities\importtools.py", line 17, in <module>
    from prefect.utilities.filesystem import filename, is_local_path, tmpchdir
  File "c:\temp\prefect\src\prefect\utilities\filesystem.py", line 5, in <module>
    import resource
ModuleNotFoundError: No module named 'resource'
```

### Versions

```Text
prefect version won't run either, but it's 2.10.17
```

### Additional context

_No response_

### LINKED ISSUE #10058: 2.10.17 failure starting prefect agent on a windows machine
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

prefect agent won't start in 2.10.17 but it does in 2.10.16

### Reproduction

```python3
(venv) C:\Users\jkang1\PycharmProjects\site_analytics\venv\Scripts>prefect agent start --pool default-agent-pool
```

### Error

```python3
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Scripts\prefect.exe\__main__.py", line 4, in <module>
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\__init__.py", line 37, in <module>
    from prefect.states import State
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\states.py", line 14, in <module>
    from prefect.client.schemas import State as State
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\client\schemas\__init__.py", line 4, in <module>
    from .actions import BlockTypeUpdate, StateCreate
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\client\schemas\actions.py", line 9, in <module>
    import prefect.client.schemas.objects as objects
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\client\schemas\objects.py", line 20, in <module>
    from prefect._internal.schemas.bases import ObjectBaseModel, PrefectBaseModel
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\_internal\schemas\bases.py", line 18, in <module>
    from prefect._internal.compatibility.experimental import experiment_enabled
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\_internal\compatibility\experimental.py", line 19, in <module>
    from prefect.settings import PREFECT_EXPERIMENTAL_WARN, SETTING_VARIABLES, Setting
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\settings.py", line 70, in <module>
    from prefect._internal.compatibility.deprecated import generate_deprecation_message
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\_internal\compatibility\deprecated.py", line 21, in <module>
    from prefect.utilities.importtools import (
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\utilities\importtools.py", line 17, in <module>
    from prefect.utilities.filesystem import filename, is_local_path, tmpchdir
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\utilities\filesystem.py", line 5, in <module>
    import resource
ModuleNotFoundError: No module named 'resource'
```

### Versions

```Text
If I try to call it I receive the same error:

(venv) C:\Users\jkang1\PycharmProjects\site_analytics\venv\Scripts>prefect version
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Scripts\prefect.exe\__main__.py", line 4, in <module>
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\__init__.py", line 37, in <module>
    from prefect.states import State
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\states.py", line 14, in <module>
    from prefect.client.schemas import State as State
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\client\schemas\__init__.py", line 4, in <module>
    from .actions import BlockTypeUpdate, StateCreate
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\client\schemas\actions.py", line 9, in <module>
    import prefect.client.schemas.objects as objects
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\client\schemas\objects.py", line 20, in <module>
    from prefect._internal.schemas.bases import ObjectBaseModel, PrefectBaseModel
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\_internal\schemas\bases.py", line 18, in <module>
    from prefect._internal.compatibility.experimental import experiment_enabled
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\_internal\compatibility\experimental.py", line 19, in <module>
    from prefect.settings import PREFECT_EXPERIMENTAL_WARN, SETTING_VARIABLES, Setting
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\settings.py", line 70, in <module>
    from prefect._internal.compatibility.deprecated import generate_deprecation_message
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\_internal\compatibility\deprecated.py", line 21, in <module>
    from prefect.utilities.importtools import (
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\utilities\importtools.py", line 17, in <module>
    from prefect.utilities.filesystem import filename, is_local_path, tmpchdir
  File "C:\Users\jkang1\PycharmProjects\site_analytics\venv\Lib\site-packages\prefect\utilities\filesystem.py", line 5, in <module>
    import resource
ModuleNotFoundError: No module named 'resource'

When I downgrade to 2.10.16 and run it:
(venv) C:\Users\jkang1\PycharmProjects\site_analytics\venv\Scripts>prefect version
Version:             2.10.16
API version:         0.8.4
Python version:      3.11.3
Git commit:          6cd7c3ee
Built:               Tue, Jun 20, 2023 2:59 PM
OS/Arch:             win32/AMD64
Profile:             remote-desktop
Server type:         cloud
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rpeden**
Prefect 2.10.17 added a Unix only import in `utilities/filesystem.py` that breaks when running Prefect server or an agent on Windows. 

Credit to @grenzi for suggesting a solution to this; this PR only differs from their PR #10055 in calling `getmaxstdio` without adding a dependency on `pywin32`.

This PR calls a function in the Windows Universal C runtime to perform the same check on Windows, and scopes the `resource` module import to ensure it won't run on Windows. Using `ctypes` in this instance should be safe, as Microsoft is unlikely to remove the Universal C runtime DLL from Windows given how many of their own applications depend on it.

See https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/getmaxstdio for more detail.

Closes #10054
Closes #10058

Although this PR adds a test for `get_open_file_limit`, many other tests in `tests/utilities/test_filesystem.py` fail on Windows due to hardcoded Unix file paths. In case it is useful, I opened a draft PR (#10065) with a potential solution that makes the failing tests work cross-platform. 

### Example
Before: run `prefect server start` on Windows and get an error: `ImportError: No module named 'resource'`
After: `prefect server start` works as expected.

The same applies for the `prefect agent start` command.

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.
  

For documentation changes:

- [ ] This pull request includes redirect settings in `netlify.toml` for files that are removed or renamed

### CODE REVIEW

**Comment by grenzi** [comment]
@rpeden  It's been many a year since i've been in the OS level api calls. 

This is great. Avoids a dependency, and uses os.name which is, from my googling the differences here on a sat morning, probably a simpler alternative. (also, today i learned about Jython, which causes os.name to return 'java' - what an odd beast it is)

Anyway, I closed my PR and referenced here since this is a better solution. Thanks!

File: src/prefect/utilities/filesystem.py [hunk 1]
```diff
@@ -122,6 +121,15 @@ def relative_path_to_current_platform(path_str: str) -> Path:
...
+
+    if os.name == "nt":
+        import ctypes
+
+        return ctypes.cdll.ucrtbase._getmaxstdio()
+    else:
+        import resource
+
+        soft_limit, _ = resource.getrlimit(resource.RLIMIT_NOFILE)
+        return soft_limit
```
Comment (desertaxle) [inline src/prefect/utilities/filesystem.py:135]: Can we catch any exceptions within this function and return a safe default (~200)?

Comment (rpeden) [reply src/prefect/utilities/filesystem.py]: @desertaxle Done, and added more tests.

**Review by desertaxle — APPROVED** [review]
LGTM! Thanks @rpeden!

**Pull request merged**